unicode.h

```diff
@@ -1,475 +1,1315 @@
+        std::string result;
+        result.reserve(codepoints.size()*4);
+        for (auto & i : codepoints) {
+            result += codepoint_to_utf8(i);
+        }
+        return result;
+    }
+
+    std::string to_string(const uint32_t & codepoints) {
+        return codepoint_to_utf8(codepoints);
+    }
+
+    bool is_category(const uint32_t & codepoint, const std::string & UNICODE_TYPE) {
+        auto it = category_name_to_category_code.find(UNICODE_TYPE);
+        if (it != category_name_to_category_code.end()) {
+            return is_category_implement(codepoint, it->second);
+        } else {
+            throw std::runtime_error("Invalid UNICODE_TYPE");
+        }
+    }
+
+    bool is_category(const uint32_t & codepoint, const std::vector<std::string> & UNICODE_TYPES) {
+        std::vector<uint32_t> category_codes;
+        category_codes.reserve(UNICODE_TYPES.size());
+
+        for (auto & UNICODE_TYPE : UNICODE_TYPES) {
+            auto it = category_name_to_category_code.find(UNICODE_TYPE);
+            if (it != category_name_to_category_code.end()) {
+                category_codes.push_back(it->second);
+            } else {
+                throw std::runtime_error("Invalid UNICODE_TYPE");
+            }
+        }
+
+        return is_category_implement_batched(codepoint, category_codes);
+    }
+
+    bool is_category(const uint32_t & codepoint, const std::vector<uint32_t> & UNICODE_TYPES) {
+        return is_category_implement_batched(codepoint, UNICODE_TYPES);
+    }
+
+    std::string to_category_name(const uint32_t & UNICODE_TYPE) {
+        auto it = category_code_to_category_name.find(UNICODE_TYPE);
+        if (it != category_code_to_category_name.end()) {
+            return it->second;
+        } else {
+            throw std::runtime_error("Invalid UNICODE_TYPE");
+        }
+    }
+
+    std::vector<std::string> to_category_code(const std::vector<uint32_t> & UNICODE_TYPES) {
+        std::vector<std::string> result;
+        result.reserve(UNICODE_TYPES.size());
+        for (auto & UNICODE_TYPE : UNICODE_TYPES) {
+            result.push_back(to_category_name(UNICODE_TYPE));
+        }
+        return result;
+    }
+
+    uint32_t to_category_code(const std::string & UNICODE_TYPE) {
+        auto it = category_name_to_category_code.find(UNICODE_TYPE);
+        if (it != category_name_to_category_code.end()) {
+            return it->second;
+        } else {
+            throw std::runtime_error("Invalid UNICODE_TYPE");
+        }
+    }
+
+    std::vector<uint32_t> to_category_code(const std::vector<std::string> & UNICODE_TYPES) {
+        std::vector<uint32_t> result;
+        result.reserve(UNICODE_TYPES.size());
+        for (auto & UNICODE_TYPE : UNICODE_TYPES) {
+            result.push_back(to_category_code(UNICODE_TYPE));
+        }
+        return result;
+    }
+
+    uint32_t get_category(const uint32_t & codepoint) {
+        return category_implement(codepoint);
+    }
+

```

Comment:
Are such indirections necessary? Just implement `get_category` - no need for private methods